File: Basics/OOPS/CPP/PaymentService.cpp
#include<iostream>
#include<map>
#include "PaymentMethod.cpp"
using namespace std;


class PaymentService{
    private: map<string, PaymentMethod*> paymentMethods;
    public: 
    PaymentService(){}
    ~PaymentService(){
        for(auto &pair : paymentMethods){
            delete pair.second;
        }
        paymentMethods.clear();
    }
    void addPaymentMethod(string &name, PaymentMethod* paymentMethod){
        paymentMethods[name] = paymentMethod;
        cout<< "Payment method added: " << name << endl;
    }
    void makePayment(const string &name, const double &amount){
        if(paymentMethods.find(name) != paymentMethods.end()) {
            paymentMethods[name]->processPayment(amount);
        } else {
            cout << "Payment method not found!" << endl;
        }
    }
};

File: Basics/OOPS/JAVA/PaymentService.java
import java.util.HashMap;
public class PaymentService{
   HashMap<String, PaymentMethod> paymentMethods;
   PaymentService(){
      paymentMethods = new HashMap<>();
   }
   public void addPaymentMethod(String name, PaymentMethod paymentMethod){
      paymentMethods.put(name, paymentMethod);
      System.out.println("Payment method added: " + name);
   }

   public void makePayment(String name, double amount){
        PaymentMethod paymentMethod = paymentMethods.get(name);
        if(paymentMethod != null){
            paymentMethod.pay(amount);
        } else {
            System.out.println("Payment method not found: " + name);
        }
   }
}